### PR TITLE
Avoid input cache when resized

### PR DESCRIPTION
**Pull request by naoyam**
C++ test copied from #27 

If the producer of a resized tensor is an input cache and we need to promote it to global memory, just do not cache the input but directly read from the global memory input. It doesn't make any sense to cache a global memory input in global memory. 

This fixes the issue of #27, which is due to the grid sync inserted for an input to pad, where the input is a copy of a fusion input.

### CODE REVIEW

File: csrc/scheduler/utils.cpp [hunk 1]
```diff
@@ -2369,23 +2370,111 @@ std::vector<TensorView*> getResizedTensors(Fusion* fusion) {
...
   auto resized_tensors = getResizedTensors(fusion);
   std::unordered_set<TensorView*> cached;
   for (auto resized_tensor : resized_tensors) {
-    for (auto producer : ir_utils::producerTvsOf(resized_tensor)) {
-      if (cached.count(producer) != 0) {
-        continue;
-      }
-      producer->cacheAfter();
-      cached.insert(producer);
+    // Resized tensors are those created by operations like pad and
```
Comment (naoyam) [inline csrc/scheduler/utils.cpp:2442]: This is an unrelated change. Instead of using `cacheAfter`, just create a copy of `producer` and use it as the input to the resize expr. 

Example:

```
tv1 = sum(tv0)
tv2 = some_resize_op(tv1);
tv3 = some_other_op(tv1);
```

When `tv1` is promoted to `Global`, we want to avoid reducing to a global memory tensor, so with `cacheAfter`:

```
tv1 = sum(tv0);
tv4 = tv1
tv4->setMemoryType(Global)
tv2 = some_resize_op(tv3)
tv3 = some_other_op(tv3);
```

This way, the reduction is done using Local, but `some_other_op` doesn't need to use the gmem copy of `tv1`. This should be just fine:

```
tv1 = sum(tv0);
tv4 = tv1
tv4->setMemoryType(Global)
tv2 = some_resize_op(tv4)
tv3 = some_other_op(tv1);
```

File: test/test_resize.cpp [hunk 2]
```diff
@@ -1591,4 +1591,73 @@ TEST_F(NVFuserTest, FusionResizeSoftmaxSliceScheduler2_CUDA) {
...
+  fusion.addOutput(tv2);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::manual_seed(0);
+
+  auto t0 = at::randn(input_shape0, options);
+  auto t1 = at::randn(input_shape1, options);
+  std::vector<c10::IValue> aten_inputs({t0, t1});
+
+  fusion.print();
```
Comment (zasdfgbnm) [inline test/test_resize.cpp:1636]: nit: remove debugging print

File: csrc/scheduler/utils.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:2441]: Is this just `cacheFork`?

File: test/test_resize.cpp [hunk 3]
```diff
@@ -323,7 +323,7 @@ TEST_F(NVFuserTest, FusionResizePad8_CUDA) {
   tv4->axis(1)->parallelize(ParallelType::TIDx);
   scheduler_utils::parallelizeAllLike(tv4);
 
-  scheduler_utils::promoteProducerMemoryTypesOfResizedTensors(&fusion);
+  scheduler_utils::promoteProducerMemoryTypesOfResizedTensors(&fusion, {});
 
   fusion.printMath();
   fusion.print();
```
[the remark this answers is not in the document]
Comment (naoyam) [reply test/test_resize.cpp:329]: Removed

File: csrc/scheduler/utils.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/scheduler/utils.cpp:2441]: I think we can extend the interface of `cacheFork` to do this?

If we have:
```C++
tv1 = sum(tv0)
tv2 = some_resize_op(tv1);
tv3 = some_other_op(tv1);
```

Can we just do 
```C++
tv1->setMemoryType(Global);
tv4 = tv1->cacheFork(/*keep_global=*/{tv2->definition()})
```

And we get:
```
tv0 -> tv4 (local) -> tv1 -> tv2
       tv4 (fork) -> tv3
```

Just some idea, have no strong opinion on having to do like this.

Comment (naoyam) [reply csrc/scheduler/utils.cpp:2441]: `cacheFork` is designed to change Fusion outputs. We could make it optional, of course. This is a simple transformation, so I don't feel it's worth consolidating into an extended `cacheFork`
